Live: adding docs (#34642)
grafana/grafana: docs/sources/live/configure-grafana-live.md

+++
title = "Configure Grafana Live"
description = "Grafana Live configuration guide"
keywords = ["Grafana", "live", "guide", "websocket"]
weight = 120
+++

# Configure Grafana Live

> Grafana Live is enabled by default. As of Grafana v8.0, it has a strict default for a maximum number of connections per Grafana server instance. 

Reviewer: ```suggestion
Grafana Live is enabled by default. In Grafana v8.0, it has a strict default for a maximum number of connections per Grafana server instance. 
```

## Max number of connections

> Grafana Live uses persistent connections (WebSocket at the moment) to deliver real-time updates to clients. You might need to tune your server infrastructure a bit.

Reviewer: ```suggestion
Grafana Live uses persistent connections (WebSocket at the moment) to deliver real-time updates to clients. Check your server infrastructure to determine whether you need to update it.
```

Author: Rewrote this part in another way.

> The number of maximum connections users can establish with Grafana is limited to 100 by default. See [max_connections]({{< relref "../administration/configuration.md#max_connections" >}}) option.

Reviewer: ```suggestion
The number of maximum connections users can establish with Grafana is limited to 100 by default. Refer to the [max_connections]({{< relref "../administration/configuration.md#max_connections" >}}) option for details. 
```

In case you want to increase this limit, ensure that your server and infrastructure allow handling more connections. The following sections discuss several common problems which could happen when managing persistent connections, in particular WebSocket connections.

### WebSocket

Grafana Live uses WebSocket protocol to deliver real-time updates to a frontend application. WebSocket is a persistent connection that starts with an HTTP Upgrade request (using the same HTTP port as the rest of Grafana) and then switches to a TCP mode where WebSocket frames can travel in both directions between a client and a server.

> Each logged-in user opens a WebSocket connection – one per browser tab.

Reviewer: ```suggestion
Each logged-in user opens a WebSocket connection, that is one per browser tab.
```

> Introducing a persistent connection leads to some things Grafana users should be aware of.

Reviewer: ```suggestion
Introducing a persistent connection can lead to certain situations that the Grafana users should be aware of.
```

Reviewer: Should we specify the situations and give a few examples?

Author: Rewrote this part in another way

### Resource usage

Each persistent connection costs some memory on a server. Typically, this should be about 50 KB per connection at this moment. Thus a server with 1 GB RAM is expected to handle about 20k connections max. Each active connection consumes additional CPU resources since the client and server send PING/PONG frames to each other to maintain a connection.

Using the streaming functionality results in additional CPU usage. The exact CPU resource utilization can be hard to estimate as it heavily depends on the Grafana Live usage pattern.

### Open file limit

Each WebSocket connection costs a file descriptor on a server machine where Grafana runs. Most operating systems have a quite low default limit for the maximum number of descriptors that process can open.

To look at the current limit on Unix run:

```
ulimit -n
```

On a Linux system, you can also check out the current limits for a running process with:

```
cat /proc/<PROCESS_PID>/limits
```

The open files limit shows approximately how many user connections your server can currently handle.
		

> 		To increase this limit, refer to [these instructions for popular operating systems](https://docs.riak.com/riak/kv/2.2.3/using/performance/open-files-limit.1.html).

Reviewer: ```suggestion
To increase this limit, refer to [these instructions](https://docs.riak.com/riak/kv/2.2.3/using/performance/open-files-limit.1.html)for popular operating systems.
```

+++
title = "Configure Grafana Live"
description = "Grafana Live configuration guide"
keywords = ["Grafana", "live", "guide", "websocket"]
weight = 120
+++

# Configure Grafana Live

Grafana Live is enabled by default. As of Grafana v8.0, it has a strict default for a maximum number of connections per Grafana server instance. 

## Max number of connections

> Grafana Live uses persistent connections (WebSocket at the moment) to deliver real-time updates to clients. You may need to tune your server infrastructure a bit.

Reviewer 2: ```suggestion
Grafana Live uses persistent connections (WebSocket at the moment) to deliver real-time updates to clients. You might need to tune your server infrastructure a bit.
```

The number of maximum connections users can establish with Grafana is limited to 100:

```

> [live]

Reviewer 2: Is this in the config file? If so, you should link to it.

Author: Fixed in e0718670d0ca16371a7a0afbf467d493448f31a0, actually I have not seen we have a detailed configuration document under administration - added Live section to it.

Using the streaming functionality results in additional CPU usage. The exact CPU resource utilization can be hard to estimate as it heavily depends on the Grafana Live usage pattern.

### Open file limit

Each WebSocket connection costs a file descriptor on a server machine where Grafana runs. Most operating systems have a quite low default limit for the maximum number of descriptors that process can open.

To look at the current limit on Unix run:

```
ulimit -n
```

On a Linux system, you can also check out the current limits for a running process with:

```
cat /proc/<PROCESS_PID>/limits
```

The open files limit shows approximately how many user connections your server can currently handle.
		

		To increase this limit, refer to [these instructions for popular operating systems](https://docs.riak.com/riak/kv/2.2.3/using/performance/open-files-limit.1.html).

### Ephemeral port exhaustion

> Ephemeral port exhaustion problem can happen between your load balancer (or reverse proxy) software and Grafana server. I.e. when you load balance requests/connections between different Grafana instances. If users connect directly to a single Grafana server instance you should not come across this issue.

Reviewer 2: ```suggestion
Ephemeral port exhaustion problem can happen between your load balancer (or reverse proxy) software and Grafana server. For example, when you load balance requests/connections between different Grafana instances. If you connect directly to a single Grafana server instance, then you should not come across this issue.
```
